### ISSUE TITLE
Possible silent out of range issue.

### ISSUE THREAD

**Issue by Sebanisu** [myst6re/deling#6]
https://github.com/myst6re/deling/blob/23aefd2306d0ed85c7d7a88831f9fe4f5dafc968/LZS.cpp#L180
The first run of `InsertNode(r);`
`r= 4098 -1; //4097`
`lson[] size is 4097;` I had changed my arrays to .at() and got the exception. Unsure what the best move is so i just increased the size of lson by 1. Also increased the size of dad because it's the same size.

**Comment by myst6re** [comment]
Hello, `r= 4098 -1; //4097` where is that? That's strange, r is initialized with the value 4078

**Comment by Sebanisu** [comment]
yeah but it's getting a minus -1 and passed to insertnode in the for loop

https://github.com/myst6re/deling/blob/23aefd2306d0ed85c7d7a88831f9fe4f5dafc968/LZS.cpp#L334

**Comment by myst6re** [comment]
I reviewed this issue, and sorry I still don't get it.

On the code you point me, you will have:

```cpp
r = 4078;
InsertNode(4077); // r - 1
InsertNode(4076); // r - 2
// [...]
InsertNode(4058); // r - 18
```

And then in InsertNode:

```cpp
p = 4097 + 0; // key[0]
if (1 >= 0) { // cmp >= 0
  if (rson[p] == 4096) {
    rson[p] = r;
    dad[r] = p;
    return;
  }
}
```

**Comment by Sebanisu** [comment]
What I did was I changed the type to std::array and used at() to access the members of lson, rson and dad. But errored for me saying out of range. Because at() checks to see if it's in range.

I could of made a mistake in my translation but I donno. I was just saying if you checked the range before using [] you might find the same error.

**Comment by Sebanisu** [comment]
Googling around there is a macro QT_CHECK_RANGE that could be defined. And I think then QArray will error if you go out of range. Though I'm unsure if this check is performed unless you use the .at(). I don't know if this info is out of date too.

I asked on discord. .at on qarray or qvector should error if the macro's are defined.
Darrell Wright(beached) — Today at 9:16 PM
```template <typename T>
inline const T &QVector<T>::at(int i) const
    { Q_ASSERT_X(i >= 0 && i < d->size, "QVector<T>::at", "index out of range");
      return d->begin()[i]; }
```
 from https://code.woboq.org/qt5/qtbase/src/corelib/tools/qvector.h.html#_ZNK7QVector2atEi
so it depends on `Q_ASSERT_X`
https://code.woboq.org/qt5/qtbase/src/corelib/global/qglobal.h.html#855
 `#if defined(QT_NO_DEBUG) && !defined(QT_FORCE_ASSERTS)` is the line that matters
old docs, I think said, it was the same as operator[]
which is at https://code.woboq.org/qt5/qtbase/src/corelib/tools/qvector.h.html#_ZNK7QVector2atEi
and it looks the same
except `operator[]` is only non-const I think
Sebanisu — Today at 9:37 PM
thanks for the info

**Comment by myst6re** [comment]
You changed the type of lson/rson or the type of the input data array? That's maybe what I don't understand from your first post

**Comment by Sebanisu** [comment]
I basically copied your code and replaced qt types with std types `std::vector` and `std::array`. Then I changed the `operator []` to `.at()`. That showed an out of range exception. Though I've made more changes since. That was my initial version of the code. In std library .at is range checked while [] is not. Qt I think is range checked if the debugging macros are enabled.

**Comment by myst6re** [comment]
For decompression I pushed a fix about that [here](https://github.com/myst6re/deling/pull/14/files?file-filters%5B%5D=.cpp#diff-ccd70e804c2f9890ae8a76fce9b4994587bd4caa8e09cac6b784a2c3371056be)

I will check the same thing for compression, I guess I will have warning since using [] on indexes out of range now generate deprecation warnings on Qt

**Comment by myst6re** [comment]
Sorry, link updated
